File: app/models/issue.py
from .db import db
from .project import Project
from .issue_type import IssueType
from .status_change import StatusChange

from datetime import datetime


class Issue(db.Model):
    __tablename__ = 'Issues'

    id = db.Column(db.Integer, primary_key=True)
    submitter_id = db.Column(db.Integer, db.ForeignKey("Users.id"), nullable=False)
    assignee_id = db.Column(db.Integer, db.ForeignKey("Users.id"))
    project_id = db.Column(db.Integer, db.ForeignKey("Projects.id"), nullable=False)
    project_idx = db.Column(db.Integer, nullable=False)
    title = db.Column(db.String(255), nullable=False)
    body = db.Column(db.String)
    type_id = db.Column(db.Integer, db.ForeignKey("IssueTypes.id"), nullable=False)
    created_at = db.Column(db.DateTime, nullable=False, default=datetime.now())
    updated_at = db.Column(db.DateTime, nullable=False, default=datetime.now())

    project = db.relationship("Project", back_populates="issues")
    comments = db.relationship("Comment", back_populates="issue")

    submitter = db.relationship("User", foreign_keys=[submitter_id], back_populates="submitted_issues")
    assignee = db.relationship("User", foreign_keys=[assignee_id], back_populates="assigned_issues")

    def to_dict(self):
        status_history = StatusChange.query.filter(StatusChange.issue_id == self.id).order_by(StatusChange.created_at.desc()).all()
        print(status_history)

        return {
            'id': self.id,
            'submitter_id': self.submitter_id,
            'assignee_id': self.assignee_id,
            'project_id': self.project_id,
            'project_idx': self.project_idx,
            'project_key': self.project.key,
            'project_personal': self.project.name == "Personal Project",
            'status_history': [status_change.to_dict() for status_change in status_history],
            'current_status': [status_change.to_dict() for status_change in status_history][0],
            'title': self.title,
            'body': self.body,
            'type_id': self.type_id,
            'comments': [comment.to_dict() for comment in self.comments],
            'submitter': self.submitter.to_dict(),
            'assignee': self.assignee.to_dict() if self.assignee else {},
            'created_at': self.created_at,
            'updated_at': self.updated_at
        }

File: react-app/src/store/issue.js
import { setUser } from './session';

// constants
const LOAD_ISSUES = 'issue/LOAD_ISSUES';
const NEW_ISSUE = 'issue/NEW_ISSUE';
const REMOVE_ISSUE = 'issue/REMOVE_ISSUE';

const loadIssues = (issues) => ({
    type: LOAD_ISSUES,
    issues,
});

// for create and edit
export const newIssue = (issue) => ({
    type: NEW_ISSUE,
    issue,
});

const removeIssue = (id) => ({
    type: REMOVE_ISSUE,
    id,
});

export const fetchIssuesByProject = (projectId) => async (dispatch) => {
    const res = await fetch(`/api/projects/${projectId}/issues/`);
    if (res.ok) {
        const data = await res.json();
        dispatch(loadIssues(data));
    }
};

export const fetchIssuesSubmittedBy = (userId) => async (dispatch) => {
    const res = await fetch(`/api/users/${userId}/submitted_issues/`);
    if (res.ok) {
        const data = await res.json();
        dispatch(loadIssues(data));
    }
};

export const fetchIssuesAssignedTo = (userId) => async (dispatch) => {
    const res = await fetch(`/api/users/${userId}/assigned_issues/`);
    if (res.ok) {
        const data = await res.json();
        dispatch(loadIssues(data));
    }
};

export const fetchIssueById = (issueId) => async (dispatch) => {
    const res = await fetch(`/api/issues/${issueId}`);
    if (res.ok) {
        const data = await res.json();
        dispatch(newIssue(data));
    }
};

export const createIssue = (issue) => async (dispatch) => {
    const issueRes = await fetch(`/api/issues/`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(issue),
    });

    if (issueRes.ok) {
        const initialData = await issueRes.json();

        const initalStatus = {
            user_id: initialData.submitter_id,
            issue_id: initialData.id,
            status_id: 1,
        };

        const statusRes = await fetch('/api/status_changes/', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(initalStatus),
        });

        if (statusRes.ok) {
            const issue = await statusRes.json();
            dispatch(newIssue(issue));
            return issue;
        } else if (statusRes.status < 500) {
            const data = await statusRes.json();
            if (data.errors) {
                return data.errors;
            }
        } else {
            return ['An error occurred. Please try again.'];
        }
    } else if (issueRes.status < 500) {
        const data = await issueRes.json();
        if (data.errors) {
            return data.errors;
        }
    } else {
        return ['An error occurred. Please try again.'];
    }
};

export const editIssue = (issue) => async (dispatch) => {
    const res = await fetch(`/api/issues/${issue.id}`, {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(issue),
    });

    if (res.ok) {
        const issue = await res.json();
        dispatch(newIssue(issue));
        return issue;
    } else if (res.status < 500) {
        const data = await res.json();
        if (data.errors) {
            return data.errors;
        }
    } else {
        return ['An error occurred. Please try again.'];
    }
};

export const deleteIssue = (issueId) => async (dispatch) => {
    const res = await fetch(`/api/issues/${issueId}`, {
        method: 'DELETE',
    });

    if (res.ok) {
        const data = await res.json();
        dispatch(setUser(data));
        dispatch(removeIssue(issueId));
    } else if (res.status < 500) {
        const data = await res.json();
        if (data.errors) {
            return data.errors;
        }
    } else {
        return ['An error occurred. Please try again.'];
    }
};

// State shape:
// state.issues --> {
//   [id]: {
//     id, submitter_id, assignee_id, project_id, project_idx, project_key, project_personal,
//     title, body, type_id, created_at, updated_at,
//     submitter: {...},
//     assignee: {...},
//     status_history: [{...}, {...}, ...],
//     current_status: {...},
//     comments: [{...}, {...}, ...],
//   },
//   [id]: {
//     id, submitter_id, assignee_id, project_id, project_idx, project_key, project_personal,
//     title, body, type_id, created_at, updated_at,
//     submitter: {...},
//     assignee: {...},
//     status_history: [{...}, {...}, ...],
//     current_status: {...},
//     comments: [{...}, {...}, ...],
//   },
// }


const initialState = {};

export default function reducer(state = initialState, action) {
    switch (action.type) {
        case LOAD_ISSUES: {
            const newState = { ...initialState };
            action.issues.forEach(issue => {
                newState[issue.id] = issue;
            });
            return newState;
        }
        case NEW_ISSUE: {
            const newState = {
                ...state,
                [action.issue.id]: action.issue,
            };
            return newState;
        }
        case REMOVE_ISSUE: {
            const newState = { ...state };
            delete newState[action.id];
            return newState;
        }
        default:
            return state;
    }
}
